### ISSUE TITLE
Batch upload tasks aren’t cancelled with DBRequestError

### ISSUE THREAD

**Issue by kfujino7** [dropbox/dropbox-sdk-obj-c#81]
I have a question regarding to the behaviour of chunked upload using batchUploadFiles.

```
 [[self restClient].filesRoutes batchUploadFiles:uploadFilesToCommitInfo queue:nil progressBlock:^(int64_t uploaded, int64_t total, int64_t expectedTotal) {
        NSLog(@"Uploaded: %lld  UploadedTotal: %lld  ExpectedToUploadTotal: %lld", uploaded, total, expectedTotal);
    } responseBlock:^(DBFILESUploadSessionFinishBatchJobStatus * result, DBASYNCPollError * routeError, DBRequestError * error) {
        if (result.isComplete) {
            NSLog(@“complete %@”, result);
        } else {
            NSLog(@"%@  %@\n", routeError, error);
       }
    }];
```

For example, if I try to upload about 45MB file, If I turn off the network after first chunk was sent, the response block caches the DBRequestError (as expected), but, as you can see below,  the task keeps trying to upload the rest of chunk 1minutes apart, after the error. It even shows the progress of uploading chunk, but it is not actually uploaded (as network is disconnected at that time).

Is it expected behaviour and do I need to cancel the tasks in the block explicitly? If so how should I implement it? 
( I tried to call `-cancel` of `DBBatchUploadTask` but it is not accessible from inside block, also tried to cancel the` [NSOperationQueue mainQueue] `but didn't work.)

> 2017-01-20 11:52:00.008 [12938:230908] Uploaded: 10485760  UploadedTotal: 10485760  ExpectedToUploadTotal: 46580427
> 
> //////  disconnect the network here //////
> 
> 2017-01-20 11:52:07.135 [12938:230908] (null)  DropboxClientError[{
>     NSError = "Error Domain=NSURLErrorDomain Code=-1009 \"The Internet connection appears to be offline.\" UserInfo={NSErrorFailingURLStringKey=https://api-content.dropbox.com/2/files/upload_session/append_v2, _kCFStreamErrorCodeKey=57, NSErrorFailingURLKey=https://api-content.dropbox.com/2/files/upload_session/append_v2, NSLocalizedDescription=The Internet connection appears to be offline., _kCFStreamErrorDomainKey=1, NSUnderlyingError=0x78602f30 {Error Domain=kCFErrorDomainCFNetwork Code=-1009 \"(null)\" UserInfo={_kCFStreamErrorDomainKey=1, NSErrorPeerAddressKey=<CFData 0x78816530 [0x1b496b8]>{length = 16, capacity = 16, bytes = 0x100201bb6ca0ac040000000000000000}, _kCFStreamErrorCodeKey=57}}}";
> }];   
> 2017-01-20 11:52:07.180 [12938:230908] Uploaded: 10485760  UploadedTotal: 20971520  ExpectedToUploadTotal: 46580427
> 2017-01-20 11:53:07.596 [12938:230908] (null)  DropboxClientError[{
>     NSError = "Error Domain=NSURLErrorDomain Code=-1001 \"The request timed out.\" UserInfo={NSErrorFailingURLStringKey=https://api-content.dropbox.com/2/files/upload_session/append_v2, _kCFStreamErrorCodeKey=-2102, NSErrorFailingURLKey=https://api-content.dropbox.com/2/files/upload_session/append_v2, NSLocalizedDescription=The request timed out., _kCFStreamErrorDomainKey=4, NSUnderlyingError=0x78932160 {Error Domain=kCFErrorDomainCFNetwork Code=-1001 \"(null)\" UserInfo={_kCFStreamErrorDomainKey=4, _kCFStreamErrorCodeKey=-2102}}}";  << (*1)
> 2017-01-20 11:53:07.599 [12938:230908] Uploaded: 10485760  UploadedTotal: 31457280  ExpectedToUploadTotal: 46580427
> 2017-01-20 11:54:08.550 [12938:230908] // same error as (*1)
> 2017-01-20 11:54:08.554 tapgroom[12938:230908] Uploaded: 10485760  UploadedTotal: 41943040  ExpectedToUploadTotal: 46580427
> 2017-01-20 11:55:08.948 [12938:230908] // same error as (*1)
> 2017-01-20 11:55:08.956 [12938:230908] Uploaded: 4637387  UploadedTotal: 46580427  ExpectedToUploadTotal: 46580427
> 2017-01-20 11:56:08.955 [12938:230908] // same error as (*1)
>

**Comment by greg-db** [comment]
Thanks for the report! I believe the expectation is that you would need to cancel this explicitly. Can you share the error(s) you got when you tried to do so? Thanks in advance!

**Comment by kfujino7** [comment]
Thank you for the reply.

If I try to cancel the task like below, 

```
__block DBBatchUploadTask* uploadTask =  [[self restClient].filesRoutes batchUploadFiles:uploadFilesToCommitInfo queue:nil progressBlock:^(int64_t uploaded, int64_t total, int64_t expectedTotal) {
        NSLog(@"Uploaded: %lld  UploadedTotal: %lld  ExpectedToUploadTotal: %lld", uploaded, total, expectedTotal);
    } responseBlock:^(DBFILESUploadSessionFinishBatchJobStatus * result, DBASYNCPollError * routeError, DBRequestError * error) {
        if (result.isComplete) {
            NSLog(@“complete %@”, result);
        } else {
            NSLog(@"%@  %@\n", routeError, error);
            [uploadTask cancel];
       }
    }];
```

I get following error

> 2017-01-23 12:56:59.366 myapp[29055:797876] *** Terminating app due to uncaught exception 'NSInternalInconsistencyException', reason: 'You must override cancel in a subclass'
> *** First throw call stack:
> (
> 	0   CoreFoundation                      0x018a1bf2 __exceptionPreprocess + 194
> 	1   libobjc.A.dylib                     0x065b3e66 objc_exception_throw + 52
> 	2   myapp                            0x00569f9b -[DBTask cancel] + 235
> 	3   myapp                            0x0056f5d3 -[DBTasksStorage cancelAllTasks] + 483
> 	4   myapp                            0x0045bf1f -[DBBatchUploadTask cancel] + 143
> 	5   myapp                            0x003612e0 __26-[DropboxUtil uploadFile:]_block_invoke.97 + 1280
> 	6   myapp                            0x0045fdb8 __145-[DBFILESRoutes(DBCustomRoutes) appendFileChunk:cursor:shouldClose:fileChunkInputStream:chunkUploadResponseQueue:chunkUploadFinished:retryCount:]_block_invoke.191 + 104
> 	7   Foundation                          0x028edf8b __NSBLOCKOPERATION_IS_CALLING_OUT_TO_A_BLOCK__ + 12
> 	8   Foundation                          0x028edc5e -[NSBlockOperation main] + 105
> 	9   Foundation                          0x028ec436 -[__NSOperationInternal _start:] + 672
> 	10  Foundation                          0x028ec18e -[NSOperation start] + 70
> 	11  Foundation                          0x028e823e __NSOQSchedule_f + 220
> 	12  libdispatch.dylib                   0x06b17cc3 _dispatch_client_callout + 14
> 	13  libdispatch.dylib                   0x06af709a _dispatch_main_queue_callback_4CF + 1755
> 	14  CoreFoundation                      0x0186205e __CFRUNLOOP_IS_SERVICING_THE_MAIN_DISPATCH_QUEUE__ + 14
> ......
> 

If I call ` [[NSOperationQueue mainQueue] cancelAllOperations];` instead of `[uploadTask cancel];`, nothing happens and following chunk upload keep happening.

**Comment by greg-db** [comment]
Thanks! We're looking into it.

**Comment by greg-db** [comment]
The 'You must override cancel in a subclass' issue in 2.0.6 is fixed now in [version 3.0.0](https://github.com/dropbox/dropbox-sdk-obj-c/releases/tag/3.0.0).

The false progress issue is still open.

**Comment by scobbe** [comment]
False progress issue is fixed now in [v3.0.3](https://github.com/dropbox/dropbox-sdk-obj-c/releases/tag/3.0.3) of the SDK. Apologies for the long delay.
